### ISSUE TITLE
Magento 2.1.15 - Invoice PDF & shipment PDF does not support Chinese Language. 

### ISSUE THREAD

**Issue by SathishDreamorbit** [magento/magento2#18260]
### Preconditions
0.  **Updated(09/19/2019):** Still actual on `2.3-develop`
1. Magento version 2.1.9
2. Magento version 2.1.15

### Steps to reproduce

1. Login to admin panel and go to Sales -> Order  
2. Search the completed order and click "view"
3. Select "Invoice" or "Shipments" tab and then click respective "Print" option.

### Expected result

1. In Invoice and Shipment the billing address should be readable.
### Actual result

1. Chinese characters are replaced by some other symbols 
![invoice_pdf](https://user-images.githubusercontent.com/41613424/46080368-6441a600-c1b7-11e8-8de3-387a1f2931b0.png)

**Comment by magento-engcom-team** [comment]
Hi @SathishDreamorbit. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@SathishDreamorbit do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-95260`, `MAGETWO-95261` were created

**Comment by dklapan** [comment]
This is a problem with fonts. The stock font is FreeSerif and it doesn't have chinese characters, you can simply edit this piece of code: 

```
    protected function _setFontRegular($object, $size = 7)
    {
        $font = \Zend_Pdf_Font::fontWithPath(
            $this->_rootDirectory->getAbsolutePath('lib/internal/GnuFreeFont/custom_font.ttf')
        );
        $object->setFont($font, $size);
        return $font;
    }
```

in `app/code/Magento/Sales/Model/Order/Pdf/AbstractPdf.php` line 856

and in lib/internal/GnuFreeFont/ you add your font with chinese characters.

**Comment by rishabhchd19** [comment]
It doesn't support Hindi either. The symbol of Rupees is displayed as []?

**Comment by SathishDreamorbit** [comment]
@Magento team,

Let me know the status of the issue.

@dklapan  I have tried the steps. However, I didn't get the respective font files for Chinese language.  Could you please share the font files if you have.

**Comment by ghost** [comment]
Hi @SathishDreamorbit You don't need this files just copy paste this -> `蔣介石蔣介石蔣介石蔣介石蔣介石蔣介石蔣介石蔣介石蔣介石蔣介石`

**Comment by SathishDreamorbit** [comment]
@engcom-backlog-nazar 
Could you please explain the steps in detail?

**Comment by ghost** [comment]
@SathishDreamorbit oh.. sorry i'm confusing a little bit, I thought you need steps to reproduce :) 
```
1 - get font with chinese characters.
2 - put in to lib/internal/GnuFreeFont/
3 - in app/code/Magento/Sales/Model/Order/Pdf/AbstractPdf.php line 856
 change $this->_rootDirectory->getAbsolutePath('lib/internal/GnuFreeFont/FreeSerif.ttf')to -> 
$this->_rootDirectory->getAbsolutePath('lib/internal/GnuFreeFont/add your font with chinese characters..ttf')
```

**Comment by SathishDreamorbit** [comment]
@engcom-backlog-nazar 

Yes, I have tried the solution. But we are unable to get the supported Chinese Characters font files. In Google, we did not receive any Chinese Font file with .ttf format. If we convert and upload it, then its not readable one and throwing an error in ZendFramework.

For Existing Magento 2 website, we cannot override framework.

**Comment by Pavel-Fekete** [comment]
Still facing this issue on 2.4-develop

**Comment by engcom-Delta** [comment]
Hi @SathishDreamorbit ,

Thanks for your reporting and collaboration.
We have re-verified the issue in latest 2.4-develop instance and the issue is reproducible.
Kindly refer the screenshots.

**Steps to reproduce -** 
1. Install Chinese Language Pack
2. Change Admin Interface to Chinese
3. Place an order successfully with Chinese address.
4. Create invoice and shipment.
5. Print the invoice and observe , Chinese characters are replaced by some other symbols
![image](https://github.com/user-attachments/assets/cc462e8d-d704-41fb-b0ad-9bd596d127ba)

Thanks.

**Comment by lbajsarowicz** [comment]
This is a known TCPDF/CJK font rendering issue in Sales PDF generation, reconfirmed by engcom-Delta on 2024-10-08. It remains unclaimed with no PR despite its age, likely because font embedding is a large-effort fix. Needs maintainer re-QA to confirm it still reproduces on current 2.4-develop before scoping the work.
